### ISSUE TITLE
Implement ModelLibrary for coron3 pipeline

### ISSUE THREAD

**Issue by stscijgbot-jp** [spacetelescope/jwst#8741]
_Issue [JP-3724](https://jira.stsci.edu/browse/JP-3724) was created on JIRA by [Ned Molter](https://jira.stsci.edu/secure/ViewProfile.jspa?name=emolter):_

In some cases, the data volume of coronagraphic observations is large enough that they cannot be processed through calwebb_coron3.  Loading all observations into memory at once should therefore be avoided if/where possible.  ModelLibrary was purpose-built to handle cases like this, and its implementation in calwebb_image3 did yield substantial memory usage improvements.  The same should be attempted for coronagraphy.

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-3724?focusedCommentId=716388#comment-716388):_

[Ned Molter](https://jira.stsci.edu/secure/ViewProfile.jspa?name=emolter) Can you give an example of some coron data that can't currently be processed through coron3?  Have they never been able to be processed?

**Comment by stscijgbot-jp** [comment]
_Comment by [Ned Molter](https://jira.stsci.edu/secure/ViewProfile.jspa?name=emolter) on [JIRA](https://jira.stsci.edu/browse/JP-3724?focusedCommentId=716393#comment-716393):_

I heard this second-hand - will update as soon as I ask some people and track down examples.

**Comment by stscijgbot-jp** [comment]
_Comment by [Ned Molter](https://jira.stsci.edu/secure/ViewProfile.jspa?name=emolter) on [JIRA](https://jira.stsci.edu/browse/JP-3724?focusedCommentId=716395#comment-716395):_

Thanks to [Hien Tran](https://jira.stsci.edu/secure/ViewProfile.jspa?name=htran) see JWSTDMS-921 linked to this ticket

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-3724?focusedCommentId=716458#comment-716458):_

Thanks, that looks like the MIRI Alpha Cen data set.  In parallel to the ModelLibrary work I'll also follow up with the MIRI coron folks about what's ideal to do with this particular data set.

**Comment by stscijgbot-jp** [comment]
_Comment by [Ned Molter](https://jira.stsci.edu/secure/ViewProfile.jspa?name=emolter) on [JIRA](https://jira.stsci.edu/browse/JP-3724?focusedCommentId=716473#comment-716473):_

Sounds good.  It looks like there are additional observations experiencing similar problems listed in the comments of that ticket.  The comment from Howard that incorrect pointings might be making resample attempt to make a huge mosaic is probably beyond the scope of what the pipeline would handle, but it initially appears that it's possible to avoid loading all science asn members into memory at once in calwebb_coron3 using ModelLibrary, so it seems worth trying out

**Comment by stscijgbot-jp** [comment]
_Comment by [Ned Molter](https://jira.stsci.edu/secure/ViewProfile.jspa?name=emolter) on [JIRA](https://jira.stsci.edu/browse/JP-3724?focusedCommentId=716658#comment-716658):_

It appears that the reason for such a high memory usage is the fact that a 4-dimensional QuadModel (which gets saved as the psfalign product) is allocated, with shape (n science integrations, n psf integrations, nrows, ncols).  This gets very large: using just 2 of the 9 input PSF CubeModels, this was already taking up 500 GB of memory.

At least at first blush it appears that this should be fixable without fundamental changes to the processing algorithms: the QuadModel is originally populated with data while looping over the zeroth axis on [this line](https://github.com/spacetelescope/jwst/blob/c2d87f7a2c8c4e383b0b4c961d802205ca0b280f/jwst/coron/imageregistration.py#L232), and then when it's accessed by klip, it's again while iterating over the zeroth axis on [this line|[https://github.com/spacetelescope/jwst/blob/c2d87f7a2c8c4e383b0b4c961d802205ca0b280f/jwst/coron/klip.py#L51].]  It therefore appears possible to keep everything stored on disk as a library of CubeModels and load them only when necessary, with data volume ~1 GB per model.

Modifying the pipeline to do this bookkeeping would be a bit tedious, and the psfalign intermediate product would be an association of CubeModels instead of a QuadModel, but it appears that substantial memory savings would be possible.

[David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) does this seem worth pursuing further?  Do you (or does anyone) know who I could ask questions about these algorithms, i.e., to make sure there's no fundamental reason the entire dataset needs to be in memory at once?

**Comment by stscijgbot-jp** [comment]
_Comment by [Jonathan Aguilar](https://jira.stsci.edu/secure/ViewProfile.jspa?name=jaguilar) on [JIRA](https://jira.stsci.edu/browse/JP-3724?focusedCommentId=716721#comment-716721):_

Your proposed solution seems reasonable to me, if that ends up being the lowest-effort solution.

Speaking for MIRI (though these probably apply to NIRCam as well), there are several strategies one could use, though I haven't had time to test them. These are in my rough order of preference:
 # Use the mean of each exposure cube (averaging across integrations) to compute the alignment, then apply the derived x/y shift to each integration within the exposure
 # Choose a single integration from each cube, and assume that JWST's pointing stability makes it so the derived x/y shift applies to all integrations within the exposure. I believe this is currently being tested on the ERS-provided spaceKLIP package <https://github.com/kammerje/spaceKLIP](https://github.com/kammerje/spaceKLIP,)>
 # Select a subset of the array to use for the alignment; for example, a square with width +/- 20 pixels on either side of the reference position in x and y. I suppose this would require a new parameter reference file to store the region. The shifts would be computed on this subset and then applied to the integration.
 # Use `cal` products, instead of `calints` products, as inputs to stage3
 ** As far as the alignment is concerned, this is equivalent to computing the mean for each exposure cube as described above
 ** This represents a tradeoff in how the wings of the PSF are modeled - greater statistical diversity but lower SNR, vs higher SNR but fewer samples. We would need to document this decision.
 ** It will affect the shape of subsequent coron3 data products.

Personally, I think that Option 1 would be the best option, as it would improve the SNR on the PSF wings and speckles that are used to compute the relative offsets. However, it would only reduce the data usage during the psf_align step, not in later steps. You would still load the full 4-dimensional cube during the `klip` stage when the PSF model is computed. Only Option 4 gets around that problem.

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-3724?focusedCommentId=716727#comment-716727):_

Thanks for the input [Jonathan Aguilar](https://jira.stsci.edu/secure/ViewProfile.jspa?name=jaguilar) .  Even in the cases that don't have an Alpha-Cen level of memory usage it sounds like there are some large performance improvements that can be realized here.  I think the best thing to do at this point is put this into the list of potential improvements to be prioritized for the build 11.2 work starting later this month (11.1 work is closing in the next 2 weeks).

Prior to that point, I'd like to tag [Julien Girard](https://jira.stsci.edu/secure/ViewProfile.jspa?name=jgirard) and [Marshall Perrin](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mperrin) for their input.  To what degree is this impacted by <https://github.com/spacetelescope/jwst/pull/8643,> which revised the coron3 alignment step to use just the first integration for alignment?

**Comment by stscijgbot-jp** [comment]
_Comment by [Marshall Perrin](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mperrin) on [JIRA](https://jira.stsci.edu/browse/JP-3724?focusedCommentId=716735#comment-716735):_

I hadn't realized that there was already the PR 8643 to revise the alignment step to only use the first integration.  Already merged to develop in fact.  I concur that's directly relevant. I think that's effectively Jonathan's option #&#x2060;2 from his list of possible strategies.  

But having not looked at the code in that PR at all myself I have to admit I'm not sure if it reduces the dimensionality sufficiently, for the reasons that Jonathan gives.

**Comment by stscijgbot-jp** [comment]
_Comment by [Ned Molter](https://jira.stsci.edu/secure/ViewProfile.jspa?name=emolter) on [JIRA](https://jira.stsci.edu/browse/JP-3724?focusedCommentId=716738#comment-716738):_

I've been looking into this a bit more this afternoon, and I don't think the way the shift is applied is directly relevant to the main memory issue.  It appears to me that the align_refs step loops over all science integrations, applying a PSF shift to each one.  Each of those 3-D shifted PSF cubes is stored in memory at the same time, as a QuadModel of shape (science_integrations, psf_integrations, rows, cols).  Then, in the klip step, this QuadModel is accessed *only* within a for loop that loops over the zeroth (science_integrations) axis.  Those shifted PSFs are then discarded and only a 3-D psf-subtracted science cube is returned.

Assuming I didn't miss some way in which the computations are inter-related, there's no reason why that 4-D array needs to be stored in memory.  Either each of the shifted PSF cubes could be stored on disk between align_refs and klip, or align_refs and klip could be combined such that the shifted PSF cube need not be stored at all.

**Comment by stscijgbot-jp** [comment]
_Comment by [Ned Molter](https://jira.stsci.edu/secure/ViewProfile.jspa?name=emolter) on [JIRA](https://jira.stsci.edu/browse/JP-3724?focusedCommentId=716817#comment-716817):_

I now realize that #8643 is indeed relevant here, because since that change, it seems that all of the planes of that 4-D PSF model are identical; that is, the same shift is applied to the same PSF for all the science integrations, and there's therefore no need to save a different PSF for each science integration at all (in memory or not).  The psfalign product should therefore be a 3-D CubeModel instead of 4-D QuadModel.  The fact that this is still being saved this way is in some ways a bug introduced by that PR, or at least an unintended consequence of that PR.

 

[Jonathan Aguilar](https://jira.stsci.edu/secure/ViewProfile.jspa?name=jaguilar) can you say more about "You would still load the full 4-dimensional cube during the `klip` stage when the PSF model is computed."?  This is not true for the current pipeline version: all processing in the klip step is done inside a `for` loop over science integrations starting on this line: <https://github.com/spacetelescope/jwst/blob/c2d87f7a2c8c4e383b0b4c961d802205ca0b280f/jwst/coron/klip.py#L42>

 

[David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) Now that we're getting to the bottom of this and realizing that the memory usage is caused by hundreds of identical copies of a shifted PSF, do you think a fix for that still needs to be put off until 11.2?  The fix would be straightforward and does not affect the output products, except that the `psfalign` intermediate product would be a CubeModel instead of a QuadModel.  And, separately, should the work remain under this ticket, since it really has nothing to do with ModelLibrary?

**Comment by stscijgbot-jp** [comment]
_Comment by [Ned Molter](https://jira.stsci.edu/secure/ViewProfile.jspa?name=emolter) on [JIRA](https://jira.stsci.edu/browse/JP-3724?focusedCommentId=716838#comment-716838):_

Thanks, I will file a separate ticket and leave this one on hold.  

 

Yes, the entire QuadModel is loaded in one go, but in one of the first lines of the for loop the QuadModel is indexed on the zeroth axis, so the processing itself occurs just on the `i`th frame of that model.  (And those are all identical)
